accepted/0000-retail-pxe-database.md

```diff
@@ -0,0 +1,259 @@
+
+Alternative is to use DNS service to lookup `tftp` CNAME, this is discussed in alternatives below.
+
+When the FQDN of `next-server` is not in the Uyuni database, such is the case when terminal is connected directly to the Uyuni, then `pxeserver_id` would be set to `null`.
+
+This is related to the [Hub Retail extension](https://github.com/uyuni-project/uyuni-rfc/pull/45) discussion about how should Retail architecture look like in Hub environment.
+
+## Storing data in database
+
+As data are provided within salt event, implementing new event handler `SaltbootPxeUpdateEvent` with accompanying `SaltbootPxeUpdateMessageAction` should be sufficient.
+
+Event handler should also schedule generation of new PXE entry.
+
+## Exposing PXE entries to users
+
+In current implementation, consumer of the `pxe_update` event data is `pxe/terminal_entry` state which is part of PXE formula.
+
+In case of keeping the final action in salt space, there is a need to provide pillar data to the state:
+
+  * by supplying directly to state when scheduling pxe generating action.
+  * by exposing database data to salt pillar system.
+
+Exposing PXE data is not hard requirement but may have some benefit like enabling user to manually apply `pxe/terminal_entry` or similar state to (re)generate pxe entries. This option can save resources by making UI/XML-RPC API changes non-essential.
+
+Database can be exposed to pillar system by external pillar mechanism connecting to the database using postgresql external pillar:
+
+```
+postgres:
+  db: susemanager
+  host: localhost
+  pass: spacewalk
+  port: 5432
+  user: spacewalk
+
+ext_pillar:
+  - postgres:
+      'retail_terminals':
+         query: "SELECT S.name, root_device, salt_device, boot_image, kernel_parameters,mac_address FROM
+                (SELECT * FROM suseRetailMinionInfo as RMI, rhnServer as S WHERE RMI.pxeserver_id = S.id AND S.name LIKE %s) AS I JOIN rhnServer AS S ON I.server_id = S.id"
+         as_list: False
+         depth: 1
+```
+
+or by writing pillar files for given PXE server using already existing Java methods provided by `MinionPillarFileManager`.
+
+
+Keeping data hidden in database has a positive side by not cluttering pillar space with rarely used data.
+To allow user to manually trigger PXE entries regeneration it is needed to develop UI or XMLRPC API endpoint, proposed below.
+
+Using these data `pxe/terminal_entry` state would need to be adapted to look for pillar data in correct places and also to support generating single pxe entry or all pxe entries depending on the values passed to the state.
+
+## Generating PXE entries
+
+Update to routine is needed to support terminals connected directly to Uyuni server. Mainly calling `pxe/terminal_entry` state using salt runner or other means locally on Uyuni.
```

Comment:
`pxe/terminal_entry` is a salt state, runner modules are different from salt states, AFAIK.

Reply:
That is true. Idea is to use [runners.salt.cmd](https://docs.saltstack.com/en/latest/ref/runners/all/salt.runners.salt.html#salt.runners.salt.cmd) to run [modules.state.single](https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.state.html#salt.modules.state.single) to apply `pxe/terminal_entry` on the master. My goal is to have the same states for both master and minion.

Another option would be to split `pxe/terminal_entry` to very simple state and execution module and then calling execution module using the same runners.salt.cmd on master.